llama-index-core/llama_index/core/workflow/checkpoint.py

```diff
@@ -0,0 +1,17 @@
+import uuid
+from llama_index.core.bridge.pydantic import (
+    BaseModel,
+    Field,
+    ConfigDict,
+)
+from typing import Optional, Dict, Any
+from .events import Event
+
+
+class Checkpoint(BaseModel):
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    id_: str = Field(default_factory=lambda: str(uuid.uuid4()))
+    last_completed_step: Optional[str]
```

Comment:
What happens when a workflow has concurrent branches? I.e. maybe a step calls send_event() and triggers multiple branches, or a map-reduce pattern

If a checkpoint is created while in this state, since we only keep track of `last_completed_step`, resuming from a checkpoint here would disregard any other concurrent parts of the workflow

Reply:
so i think i've got this covered now. See the updated Parallel Executions notebook, but the jist of it is I needed to add a `in_progress` steps to `Context`. If there are any steps in progress during a checkpoint creation time, then when loading up that checkpoint, those in progress steps will need to be started all over again.

Reply:
I think one ambiguity (that is maybe solved and I missed it) -- each `workflow.run()` generates a run ID -- but how do I go from this

```
result = await workflow.run(..., store_checkpoints=True)
```

To getting the run ID for that specific run?

Reply:
I guess if you log the current run ids before and after, you should know (assuming no other runs were started in the meantime)

Reply:
do you think maybe we could add it to the `handler` perhaps through some field like `last_run_id` or something.

Reply:
we can probably just directly add a `run_id` to `WorkflowHanlder`, that feels like a nice way to access this

Reply:
agreed! Added `run_id` attribute to `WorkflowHandler` -- the `__init__` accepts an optional param default to None.